### PR TITLE
feat: batch deposit&withdraw workflow

### PR DESCRIPTION
**Pull request by ShookLyngs**
### Description
This PR provides a workflow to run on GitHub Actions (triggered manually).
The workflow is about `Batch deposit & withdraw`, in which we will try to make several rounds of deposit and withdraw for a group of accounts in a short period of time, to see if there will be problems with Godwoken on alphanet_v1.

### Changes
- Account-Faucet (the CLI tool to claim faucet)
  - Refactor account-faucet with clearer commands
  - Add batch relevant commands
- Light-Godwoken-CLI
  - Provide clearer console messages
  - Add batch deposit&withdraw relevant commands
- GitHub Workflows
  - Add `batch-prepare` workflow to claim faucet for derived accounts in batch actions
  - Add `batch-deposit-withdraw` for long-running testing on alphanet_v1

### Details

How to get started:
1. Run the `batch-prepare` workflow and it should automatically claim L1&L2 capacity for the derived accounts, and transfer test sUDT from a default issuer account to the derived accounts
2. Run the `batch-deposit-withdraw` workflow and it should keep sending batch deposit and withdrawal transactions whenever the Godwoken block number changes (For now the workflow will run for 180 minutes, later we shall make it run for 240 minutes, and start a new workflow every 240 minutes)

Terms:
- Derived accounts - Derived accounts are generated from the main account we provided, for now the main account is `secrets.GODWOKEN_TEST_PRIVATE_KEY`

### TODO:

batch-prepare
- [x] Make it claim L1&L2 capacity for the derived accounts
- [x] Transfer sUDT to L1&L2 of the derived accounts
- [x] Generate and upload artifacts with action records for the workflow

batch-deposit-withdraw
- [x] Send a batch of deposit&withdraw requests for each Godwoken `blockNumber`
- [x] Generate and upload artifacts with action records for the workflow
- [x] There are some Uncaught Exceptions keep crashing the Node process, trying to fix it
- [x] Schedule the workflow to run every 4 hours (because there will be delay for each workflow, it can be hard to make each workflow end before the 6 hour limit)

### CODE REVIEW

File: scripts/account-faucet/src/commands/batch-accounts-l2.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    )
+    .action(batchAccountsForL2)
+  ;
+}
+
+export async function batchAccountsForL2(params: {
+  privateKey: HexString;
+  derivedCount: string;
+  network: Network;
+  ckbAddress?: Address;
```
Comment (Flouse) [inline scripts/account-faucet/src/commands/batch-accounts-l2.ts]: unused parm?

File: .github/workflows/account-faucet.yml [hunk 2]
```diff
@@ -2,7 +2,7 @@ name: Account Faucet
 
 on:
   schedule:
-  - cron: '0 0 * * *' # every 24 hours
+  - cron: '0 6 * * *' # everyday at 06:00 AM (TODO: what timezone?)
```
Comment (Flouse) [inline .github/workflows/account-faucet.yml]: I think the timezone of GitHub Action is UTC Time T0.

**Review by Flouse — CHANGES_REQUESTED** [review]
```bash
lgc get-balance 0x***** -n testnet_v1

(node:20215) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
(Use `node --trace-deprecation ...` to show where the warning was created)
l1-address: ckb1qz0n46hjl3pe2jwtepcvv5ehf9p6l94qvk9addgmarvfsvvrum6j7qgpje4npetk5nt8xxvkwj953ht8e9809yr8qq0h6x88
l1-balance: 0.0
l2-address: 0x966B30e576A4d6731996748B48Dd67C94eF29067
l2-balance: 592231.102233887316314413
```

I got a wrong l1-address. `ckb...` is mainnet address.
Not urgent, maybe we can check and fix it later.

File: scripts/light-godwoken-cli/src/utils/format.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
+import { HexString } from '@ckb-lumos/base';
+
+export function isAllDefinedOrAllNot(targets: unknown[]) {
```
Comment (Flouse) [inline scripts/light-godwoken-cli/src/utils/format.ts:3]:
```suggestion
export function isAllDefined(targets: unknown[]) {
```

**Review by Flouse — CHANGES_REQUESTED** [review]

File: scripts/light-godwoken-cli/package.json [hunk 4]
```diff
@@ -1,30 +1,34 @@
...
     "esbuild": "^0.15.3",
-    "ts-node": "^10.4.0",
-    "typescript": "^4.5.4"
+    "keccak256": "^1.0.6",
+    "ts-node": "^10.9.1"
   },
   "dependencies": {
     "@ckb-lumos/base": "^0.18.0",
+    "@ckitjs/ckit": "^0.2.0",
+    "@polyjuice-provider/base": "^0.1.5",
```
Comment (Flouse) [inline scripts/light-godwoken-cli/package.json]: https://github.com/nervosnetwork/polyjuice-provider has been deprecated, please avoid using it if it is possible.

**Review by Flouse — CHANGES_REQUESTED** [review]

File: scripts/light-godwoken-cli/src/utils/format.ts (hunk 3 above)
Comment (ShookLyngs) [reply scripts/light-godwoken-cli/src/utils/format.ts:3]: I double checked the function and the logic of the function does not cover exactly what I want. 
I'll fixed it later.

File: scripts/light-godwoken-cli/package.json (hunk 4 above)
Comment (ShookLyngs) [reply scripts/light-godwoken-cli/package.json]: Okay, I used this library to help me get `sudt_id` from a `sudt_type_args`, any better replacement?

File: scripts/account-faucet/src/commands/batch-accounts-l2.ts (hunk 1 above)
Comment (ShookLyngs) [reply scripts/account-faucet/src/commands/batch-accounts-l2.ts]: Okay, and maybe I should combine `batch-accounts-l1` and `batch-accounts-l2` into a `get-batch-accounts` command.

**Comment by ShookLyngs** [comment]
> I got a wrong l1-address. `ckb...` is mainnet address. Not urgent, maybe we can check and fix it later.

This is a bug in branch `ref-custom-tokens` of light-godwoken repo due to incorrectly swapping the mainnet and testnet configs. I have fixed it locally, and maybe I will push it to remote tomorrow, along with other fixes.

File: scripts/light-godwoken-cli/package.json (hunk 4 above)
Comment (Flouse) [reply scripts/light-godwoken-cli/package.json]: I think you could just add `getAccountIdByScriptHash` method into LightGodwokenProvider interface ?

File: scripts/account-faucet/src/commands/batch-claim-l2.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  program
+    .command('batch-claim-l2')
+    .description('claim faucet for L2 derived accounts based on a private-key')
+    .requiredOption('-p, --private-key <HEX_STRING>', 'ckb private key')
+    .option('-c, --derived-count <INT>', 'amount of derived accounts to use', '30')
+    .option('-c --ckb-address <ADDRESS>', 'ckb deposit-from address')
+    .addOption(
+      new Option('-n, --network <NETWORK>', 'network to use')
+        .choices(Object.values(Network))
+        .default(Network.TestnetV1)
```
Comment (classicalliu) [inline scripts/account-faucet/src/commands/batch-claim-l2.ts:18]: Some commands defaults to alphanet and some to testnet ?

File: scripts/account-faucet/src/faucet/address.ts [hunk 6]
```diff
@@ -32,38 +35,51 @@ export async function encodeLayer2DepositAddress(config: NetworkConfig, gw: Godw
...
+  privateKey = toHexString(privateKey);
+  return ethersUtils.computeAddress(privateKey);
 }
 
 export function toHexString(target: string): HexString {
   return target.startsWith('0x') ? target : `0x${target}`;
 }
 
-export function toAddress(target: string): Address {
+export function toNonHexString(target: string): string {
```
Comment (classicalliu) [inline scripts/account-faucet/src/faucet/address.ts]: `toHexStringWithoutPrefix` or `removeHexStringPrefix` maybe clearer ?

File: scripts/account-faucet/src/faucet/derived.ts [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+import { utils } from 'ethers';
+
+export interface DerivedAccount {
+  privateKey: HexString;
+  ethAddress: Address;
+}
+export function privateKeyToDerivedAccounts(privateKey: HexString, count: number): DerivedAccount[] {
+  const keys: HexString[] = [privateKey];
+  for (let i = 0; i < count; i++) {
+    const newKey = '0x' + keccak256(keys[keys.length - 1]).toString('hex');
```
Comment (classicalliu) [inline scripts/account-faucet/src/faucet/derived.ts]: secp256k1 private key has a max value, not sure the key generated here is always a valid key.

File: scripts/light-godwoken-cli/package.json [hunk 8]
```diff
@@ -1,30 +1,34 @@
...
   },
   "devDependencies": {
     "@ckb-lumos/config-manager": "^0.18.0",
     "@ckb-lumos/lumos": "^0.18.0",
     "@ethersproject/abstract-signer": "^5.6.2",
     "esbuild": "^0.15.3",
-    "ts-node": "^10.4.0",
-    "typescript": "^4.5.4"
+    "keccak256": "^1.0.6",
+    "ts-node": "^10.9.1"
```
Comment (classicalliu) [inline scripts/light-godwoken-cli/package.json]: Why to remove `typescript` ?

File: scripts/account-faucet/src/faucet/derived.ts (hunk 7 above)
Comment (ShookLyngs) [reply scripts/account-faucet/src/faucet/derived.ts]: You're right, is it enough to set the max count to `64`?

Comment (classicalliu) [reply scripts/account-faucet/src/faucet/derived.ts]: It seems not related to max try count, a new key is a hash of previous key, so the key looks not incrementing.

Comment (ShookLyngs) [reply scripts/account-faucet/src/faucet/derived.ts]: I've added TODOs to derived accounts relevant methods, this part of content will be refactored with hd-wallet.

**Comment by ShookLyngs** [comment]
After running in my repo for a long time, the workflow is now stable.
Although the derived account generation part still has a problem, but for now the whole workflow should work, so I thinnk we can fix the derived account generation issue later.

**Review by Flouse — APPROVED** [review]

**Pull request merged**